File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans=[]
        candidates.sort()
        def rec(start,curr_sum,temp):
            if curr_sum>target or start>len(candidates):
                return
            if curr_sum==target:
                ans.append(temp[:])
                return
            for i in range(start,len(candidates)):
                if i>start and candidates[i]==candidates[i-1]:
                    continue
                temp.append(candidates[i])
                rec(i+1,curr_sum+candidates[i],temp)
                temp.pop()
        rec(0,0,[])
        return ans

File: Data Structures & Algorithms/generate-parentheses/submission-0.java
class Solution {
    public void helper(List<String> ans, int o,int c, int n,String temp){
        if(o>n) return;
        if(c>n) return;
        if(o+c==2*n){
            ans.add(temp);
            return;
        }
        if(o<=c){
            temp+="(";
            helper(ans,o+1,c,n,temp);
        }else{
            helper(ans,o+1,c,n,temp+"(");
            helper(ans,o,c+1,n,temp+")");
            
        }
    }
    public List<String> generateParenthesis(int n) {
        int o=0;
        int c=0;
        List<String> ans=new ArrayList<>();
        helper(ans,o,c,n,"");
        return ans;

    }
}
